lib/manager/npm/extract/monorepo.ts

```diff
@@ -13,7 +20,115 @@ function matchesAnyPattern(val: string, patterns: string[]): boolean {
   return res;
 }
 
-export function detectMonorepos(packageFiles: Partial<PackageFile>[]): void {
+export async function extractPnpmFilters(
+  fileName: string
+): Promise<string[] | null> {
+  try {
+    const contents = yaml.safeLoad(await readLocalFile(fileName, 'utf8'), {
+      json: true,
+    }) as any;
+    if (
+      !Array.isArray(contents?.packages) ||
+      !contents.packages.every((item) => typeof item === 'string')
+    ) {
+      logger.debug(
+        { fileName },
+        'Failed to find required "packages" array in pnpm-workspace.yaml'
+      );
+      return null;
+    }
+    return contents.packages;
+  } catch (error) {
+    logger.debug({ fileName }, 'Failed to parse pnpm-workspace.yaml');
+    return null;
+  }
+}
+
+export async function findPnpmWorkspace(
+  packageFile: string,
+  normalizedLocalDir: string
+): Promise<{ lockFilePath: string; workspaceYamlPath: string } | null> {
+  const workspaceYamlPath = await findUp('pnpm-workspace.yaml', {
+    cwd: upath.dirname(upath.join(normalizedLocalDir, packageFile)),
+    type: 'file',
+  });
+  if (
+    !workspaceYamlPath ||
+    !upath.normalizeSafe(workspaceYamlPath).startsWith(normalizedLocalDir)
```

Comment:
I cribbed this from https://github.com/renovatebot/renovate/blob/8b0f0ab8c233903fb810675898a8630dd8939973/lib/manager/nuget/util.ts#L52

I may be missing some details around how "localDir" interacts with the fs utils.

Reply:
I think we might want to centralize this into our `util/fs` if it's used by more than one. The idea is to "find up" without exiting the `localDir` (which is the repo)